orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -825,6 +819,208 @@ void AddNewScalarNodeArgAndInitializer(Graph& graph,
+  node_queue.push(startNode);
+
+  while (!node_queue.empty()) {
+    auto node = node_queue.front();
+    node_queue.pop();
+    if (visited_nodes.count(node) == 0) {
+      visited_nodes.insert(node);
+      std::vector<const Node*> connected_nodes;
+      ORT_THROW_IF_ERROR(FindAllConnectedNodes(graph, node, connected_nodes, visited_inputs, visited_outputs));
+
+      for (auto n : connected_nodes) {
+        ORT_ENFORCE(n != nullptr, "Found nullptr in searching for connected nodes");
+        node_queue.push(n);
+      }
+    }
+  }
+}
+
+// If an initializer is shared across partitions, instead of creating a separate all_reduce op to
+// sync with those tensors in selected partitions, we save only one copy of that initializer in
+// the very first partition it appears, and pass that data down to all following partitions
+// where this initializer is used.
+common::Status HandleSharedInitializer(Graph& graph,
+                                       const std::vector<Node*>& send_nodes,
+                                       const std::vector<Node*>& recv_nodes) {
+  // Map an given initializer to all the partitions that its consumer nodes reside. The size of
+  // the mapped vector reflects how many partitions this initializer's consumer nodes distribute.
+  // If its size is greater than 1, it means this initializer is being used in more than one partition and
+  // we need to proceed those cases.
+  std::map<const NodeArg*, std::vector<PipelineStageNodeGroup>> input_consumer_stage_map;
+
+  for (size_t stage = 0; stage <= send_nodes.size(); ++stage) {
+    std::set<const Node*> visited_nodes;
+    std::set<const NodeArg*> visited_inputs;
+    std::set<const NodeArg*> visited_outputs;
+
+    if (stage < send_nodes.size()) {
+      TraverseGraphWithConnectedElement(graph, send_nodes[stage],
+                                        visited_nodes, visited_inputs, visited_outputs);
+    } else {
+      TraverseGraphWithConnectedElement(graph, recv_nodes.back(),
+                                        visited_nodes, visited_inputs, visited_outputs);
+    }
+
+    for (const auto input : visited_inputs) {
+      // If the node is an input instead of an initializer, continue
+      if (!graph.IsInitializerTensor(input->Name())){
+        continue;
+      }
+
+      // group all consumer nodes that shares the same input initializer in visited_consumer_nodes
+      std::vector<Node*> consumer_nodes = graph.GetMutableConsumerNodes(input->Name());
+      std::vector<Node*> visited_consumer_nodes;
+      for(auto consumer_node : consumer_nodes){
+        if (visited_nodes.count(consumer_node) != 0){
+          visited_consumer_nodes.push_back(consumer_node);
+        }
+      }
+
+      if (input_consumer_stage_map.count(input) == 0) {
+        std::vector<PipelineStageNodeGroup> stage_node_group{PipelineStageNodeGroup(stage, visited_consumer_nodes)};
+        input_consumer_stage_map[input] = std::move(stage_node_group);
```

Comment:
Would it work to do this in a single line:

```
input_consumer_stage_map[input] = {PipelineStageNodeGroup(stage, visited_consumer_nodes};
```

...or is there a C++ type inference problem that prevents this? :)

Reply:
yeah, the initializer list failed on type inferencing. I can change to create an in-place vector though.

---
In reply to: [449306300](https://github.com/microsoft/onnxruntime/pull/4321#discussion_r449306300) [](ancestors = 449306300)